### PR TITLE
8311177: Switching to interpreter only mode in carrier thread can lead to crashes

### PR DESCRIPTION
**Pull request by sspitsyn**
Please, review the following `interp-only` issue related to carrier threads.
There are 3 problems fixed here:
 - The  `EnterInterpOnlyModeClosure::do_threads` is taking the `JvmtiThreadState` with the `jt->jvmti_thread_state()` which is incorrect when we have a deal with a carrier thread. The target state is known at the point when the `HandshakeClosure` is set, so the fix is to pass it as a constructor parameter.
 - The `state->is_pending_interp_only_mode())` was processed at mounts only but it has to be processed for unmounts as well. 
 - The test `test/hotspot/jtreg/serviceability/jvmti/vthread/MethodExitTest/libMethodExitTest.cpp` has a wrong assumption that there can't be `MethodExit` event on the carrier thread when the function `breakpoint_hit1` is being executed. However, it can happen if the virtual thread gets unmounted.
 
 The fix also includes new test case `vthread/CarrierThreadEventNotification` developed by Patricio.
 
 Testing:
 - Ran new test case locally
 - Ran mach5 tiers 1-6

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311177](https://bugs.openjdk.org/browse/JDK-8311177): Switching to interpreter only mode in carrier thread can lead to crashes (**Bug** - P3)

### Reviewers
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**)
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19438/head:pull/19438` \
`$ git checkout pull/19438`

Update a local copy of the PR: \
`$ git checkout pull/19438` \
`$ git pull https://git.openjdk.org/jdk.git pull/19438/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19438`

View PR using the GUI difftool: \
`$ git pr show -t 19438`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19438.diff">https://git.openjdk.org/jdk/pull/19438.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19438#issuecomment-2136335209)

### CODE REVIEW

File: test/hotspot/jtreg/serviceability/jvmti/vthread/MethodExitTest/libMethodExitTest.cpp [hunk 1]
```diff
@@ -197,12 +197,8 @@ breakpoint_hit2(jvmtiEnv *jvmti, JNIEnv* jni,
...
   jvmtiError err;
 
-  // Verify that we did not get a METHOD_EXIT events when enabled on the cthread.
-  if (received_method_exit_event) {
-    passed = JNI_FALSE;
-    received_method_exit_event = JNI_FALSE;
-    LOG("FAILED: got METHOD_EXIT event on the cthread: %p\n", cthread);
-  }
+  // need to reset this value after the breakpoint_hit1
+  received_method_exit_event = JNI_FALSE;
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/MethodExitTest/libMethodExitTest.cpp:201]: There was a loom-dev email thread regarding this last year. Seems related. I had concluded that the way the test was written that no MethodExit event should have been received. I'm not sure if I missed something in my analysis or if this failure is a result of your changes:

https://mail.openjdk.org/pipermail/loom-dev/2023-August/006059.html
https://mail.openjdk.org/pipermail/loom-dev/2023-September/006170.html

File: test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+static jthread* tested_cthreads = nullptr;
+static jint cthread_cnt = 0;
+
+static const char* CTHREAD_NAME_START = "ForkJoinPool";
+static const int CTHREAD_NAME_START_LEN = (int)strlen("ForkJoinPool");
+
+static jint
+get_cthreads(JNIEnv* jni, jthread** cthreads_p) {
+  jthread* tested_cthreads = NULL;
```
Comment (alexmenkov) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp:44]:
```suggestion
  jthread* tested_cthreads = nullptr;
```

Comment (alexmenkov) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp:44]: This local variable has the same name as global.
I'd suggest to rename the local var or remove it (and the function should set both `tested_cthreads` and ` cthread_cnt`)

File: test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+      jvmtiError err = jvmti->SetEventNotificationMode(JVMTI_ENABLE, JVMTI_EVENT_SINGLE_STEP, thread);
+      check_jvmti_status(jni, err, "event handler: error in JVMTI SetEventNotificationMode for event JVMTI_EVENT_SINGLE_STEP");
+    }
+  } else {
+    if (tested_cthreads == nullptr) {
+      jni->FatalError("Should be set");
+    }
+    for (int i = 0; i < cthread_cnt; i++) {
+      jthread thread = tested_cthreads[i];
+      char* tname = get_thread_name(jvmti, jni, thread);
```
Comment (alexmenkov) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp:91]: `tname` is not needed

File: test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/CarrierThreadEventNotification.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/CarrierThreadEventNotification.java:2]: Fixed, thanks.

File: test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp:2]: Fixed, thanks.

File: test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+extern "C" {
+
+// set by Agent_OnLoad
+static jvmtiEnv* jvmti = nullptr;
+
+static jthread* tested_cthreads = nullptr;
+static jint cthread_cnt = 0;
+
+static const char* CTHREAD_NAME_START = "ForkJoinPool";
+static const int CTHREAD_NAME_START_LEN = (int)strlen("ForkJoinPool");
```
[the remark this answers is not in the document]
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp:40]: Fixed, thanks.

File: test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp (hunk 2 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp:44]: Fixed, thanks.

Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp:44]: Thanks. Renamed the local to `cthreads` and the global to `carrier_threads`.

File: test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp (hunk 3 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp:91]: Removed, thanks.

File: test/hotspot/jtreg/serviceability/jvmti/vthread/MethodExitTest/libMethodExitTest.cpp (hunk 1 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/MethodExitTest/libMethodExitTest.cpp:201]: Thank you for the comment and links to the discussion. In fact, I've observed the MethodExit events really posted between the breakpoint hits: `hit1` and `hit2`. The first one is at the return from the `unmount()` method. I was not able to prove why they should not be expected.

**Review by pchilano — COMMENTED** [review]
Hi Serguei,

Thanks for fixing this one.

File: src/hotspot/share/prims/jvmtiThreadState.cpp [hunk 7]
```diff
@@ -663,6 +670,8 @@ JvmtiVTMSTransitionDisabler::VTMS_unmount_begin(jobject vthread, bool last_unmou
   if (!last_unmount) {
     thread->rebind_to_jvmti_thread_state_of(thread->threadObj());
   }
+  // enable interp_only_mode for carrier thread if it has pending bit
+  process_pending_interp_only(thread);
```
[the remark this answers is not in the document]
Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiThreadState.cpp]: Thank you for the comment!
I was also thinking about placing it to the `rebind_to_jvmti_thread_state_of()`. 
I've made  and pushed this change now.

File: test/hotspot/jtreg/serviceability/jvmti/vthread/MethodExitTest/libMethodExitTest.cpp (hunk 1 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/MethodExitTest/libMethodExitTest.cpp:201]: I'm not sure I follow the test logic. Its summary says "Verifies that MethodExit events are delivered on both carrier and virtual threads", but now it just ignores MethodExit requested for carrier thread in breakpoint_hit1.
Then there is no sense to request the event on carrier thread.
Per the test summary I'd expect the test should test MethodExit for carrier thread, but then java part needs to force unmount

File: test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+extern "C" {
+
+// set by Agent_OnLoad
+static jvmtiEnv* jvmti = nullptr;
+
+static jthread* carrier_threads = nullptr;
+static jint cthread_cnt = 0;
+
+static const char* CTHREAD_NAME_START = "ForkJoinPool";
+static const size_t CTHREAD_NAME_START_LEN = (int)strlen("ForkJoinPool");
```
[the remark this answers is not in the document]
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp:40]: Thanks, fixed now.

File: test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+  check_jvmti_status(jni, err, "get_cthreads: error in JVMTI GetAllThreads");
+
+  for (int idx = 0; idx < all_cnt; idx++) {
+    jthread thread = cthreads[idx];
+    char* tname = get_thread_name(jvmti, jni, thread);
+
+    if (strncmp(tname, CTHREAD_NAME_START, CTHREAD_NAME_START_LEN) == 0) {
+      cthreads[ct_cnt++] = jni->NewGlobalRef(thread);
+    }
+    deallocate(jvmti, jni, (void*)tname);
```
[the remark this answers is not in the document]
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp:58]: Why do you think, it is not needed?

This is the `deallocate()` function in the `jvmti_common.hpp`:
```
static void
deallocate(jvmtiEnv *jvmti, JNIEnv* jni, void* ptr) {
  jvmtiError err = jvmti->Deallocate((unsigned char*)ptr);
  check_jvmti_status(jni, err, "deallocate: error in JVMTI Deallocate call");
}
```

File: test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+      jni->FatalError("Should be set");
+    }
+    for (int i = 0; i < cthread_cnt; i++) {
+      jthread thread = carrier_threads[i];
+      jvmtiError err = jvmti->SetEventNotificationMode(JVMTI_DISABLE, JVMTI_EVENT_SINGLE_STEP, thread);
+      check_jvmti_status(jni, err, "event handler: error in JVMTI SetEventNotificationMode for event JVMTI_EVENT_SINGLE_STEP");
+      jni->DeleteGlobalRef(thread);
+    }
+    jvmtiError err = jvmti->Deallocate((unsigned char*)carrier_threads);
+    check_jvmti_status(jni, err, "deallocate: error in JVMTI Deallocate call");
```
[the remark this answers is not in the document]
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/CarrierThreadEventNotification/libCarrierThreadEventNotification.cpp:96]: Thanks, fixed now.

**Comment by sspitsyn** [comment]
> I'm not sure I follow the test logic. Its summary says "Verifies that MethodExit events are delivered on both carrier and virtual threads", but now it just ignores MethodExit requested for carrier thread in breakpoint_hit1.
Then there is no sense to request the event on carrier thread.
> Per the test summary I'd expect the test should test MethodExit for carrier thread, but then java part needs to force unmount.

As we already agreed I've filed the cleanup test bug:
[8333459](https://bugs.openjdk.org/browse/JDK-8333459) cleanup and check MethodExit events are posted on carrier threads in MethodExitTest

**Review by alexmenkov — APPROVED** [review]

**Comment by sspitsyn** [comment]
Thank you for review, Alex!

**Review by pchilano — APPROVED** [review]
Thanks Serguei, looks good to me.

**Comment by sspitsyn** [comment]
Thank you for review, Patricio!

**Pull request closed without merging**